athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsMetadataHandler.java

```diff
@@ -0,0 +1,471 @@
+                tableInfo.getSchemaName(), tableInfo.getTableName());
+        datasource.loadAllTables(tableInfo.getSchemaName());
+        Optional<StorageTable> optionalTable = datasource.getStorageTable(tableInfo.getSchemaName(),
+                tableInfo.getTableName());
+        if (optionalTable.isPresent()) {
+            StorageTable table = optionalTable.get();
+            bucketName = table.getParameters().get(TABLE_PARAM_BUCKET_NAME);
+            objectName = table.getParameters().get(TABLE_PARAM_OBJECT_NAME);
+        }
+
+        List<StorageSplit> storageSplits;
+        try {
+            storageSplits = datasource.getStorageSplits(request.getSchema(), request.getConstraints(),
+                    request.getTableName(), bucketName, objectName);
+            requireNonNull(storageSplits, "Storage split is null in GcsMetadataHandler:doGetSplits");
+        }
+        catch (IOException exception) {
+            throw new RuntimeException("Exception occurred: " + exception.getMessage(), exception);
+        }
+        requireNonNull(storageSplits, "List of splits can't be retrieve from metadata");
+
+        Block partitions = request.getPartitions();
+        LOGGER.info("MetadataHandler=GcsMetadataHandler|Method=doGetSplits|Message=Partition block {}", partitions);
+        LOGGER.info("MetadataHandler=GcsMetadataHandler|Method=doGetSplits|Message=Block partition row count {}",
+                partitions.getRowCount());
+        Set<Split> splits = new HashSet<>();
+        final ObjectMapper objectMapper = new ObjectMapper();
+        if (partitions.getRowCount() == 0) {
+            return getSingleSplit(request, bucketName, objectName, storageSplits, splits, objectMapper);
+        }
+
+        int partitionContd = decodeContinuationToken(request);
+        List<Integer> storageSplitListIndices = getSplitIndices(partitions);
+        LOGGER.info("MetadataHandler=GcsMetadataHandler|Method=doGetSplits|Message=Start splitting from position {}",
+                partitionContd);
+        if (storageSplitListIndices.isEmpty()) {
+            LOGGER.debug("No more storage split indices, returning empty split with null continuation token");
+            return new GetSplitsResponse(request.getCatalogName(), splits, null);
+        }
+        int startSplitIndex = storageSplitListIndices.get(0);
+        LOGGER.info("Current split start index {}", startSplitIndex);
+        for (int curPartition = partitionContd; curPartition < partitions.getRowCount(); curPartition++) {
+            int currentSplitIndex = startSplitIndex + curPartition;
+            SpillLocation spillLocation = makeSpillLocation(request);
+            StorageSplit storageSplit = storageSplits.get(currentSplitIndex);
+            String storageSplitJson;
+            try {
+                storageSplitJson = GcsParquetSplitUtil.splitAsJson(storageSplit);
```

Comment:
This is parquet specific?

Reply:
I moved it to com.amazonaws.athena.connectors.gcs.GcsUtil